superduperdb/ext/torch/model.py

```diff
@@ -94,45 +93,152 @@ def __post_init__(self):
+        return self.object.state_dict()
+
+    @contextmanager
+    def saving(self):
+        with super().saving():
+            was_training = self.object.training
+            try:
+                self.object.eval()
+                yield
+            finally:
+                if was_training:
+                    self.object.train()
 
-        def _validate(
-            self, db: Datalayer, validation_set: Dataset, metrics: t.Sequence[Metric]
+    def __getstate__(self):
+        state = self.__dict__.copy()
+        if isinstance(self.object, torch.jit.ScriptModule) or isinstance(
+            self.object, torch.jit.ScriptFunction
         ):
-            raise NotImplementedError
+            f = io.BytesIO()
+            torch.jit.save(self.object, f)
+            state['_object_bytes'] = f.getvalue()
+        return state
 
-        def append_metrics(self, d: t.Dict[str, float]) -> None:
-            raise NotImplementedError
+    def __setstate__(self, state):
+        keys = state.keys()
+        for k in keys:
+            if k != '_object_bytes':
+                self.__dict__[k] = state[k]
+            else:
+                state.__dict__['object'] = torch.jit.load(
+                    io.BytesIO(state.pop('object_bytes'))
+                )
 
-    @property
-    def optimizers(self) -> t.List:
-        raise NotImplementedError
+    def _predict_one(self, x):
+        with torch.no_grad(), eval(self.object.artifact):
+            if self.preprocess is not None:
+                x = self.preprocess.artifact(x)
+            x = to_device(x, device_of(self.object.artifact))
+            singleton_batch = create_batch(x)
+            method = getattr(self.object.artifact, self.forward_method)
+            output = method(singleton_batch)
+            output = to_device(output, 'cpu')
+            args = unpack_batch(output)[0]
+            if self.postprocess is not None:
+                args = self.postprocess.artifact(args)
+            return args
 
-    def preprocess(self, r):
-        raise NotImplementedError  # implemented in PyTorch wrapper and PyTorch pipeline
+    def _predict(self, x, one: bool = False, **kwargs):  # type: ignore[override]
+        with torch.no_grad(), eval(self.object.artifact):
+            if one:
+                return self._predict_one(x)
+            if not isinstance(self.preprocess, Artifact):
 
-    def save(self, database: Datalayer):
-        raise NotImplementedError
+                def func(x):
+                    return x
+
+            else:
+                func = self.preprocess.artifact
+            inputs = BasicDataset(x, func)
+            loader = torch.utils.data.DataLoader(inputs, **kwargs)
+            out = []
+            for batch in tqdm(loader, total=len(loader)):
+                batch = to_device(batch, device_of(self.object.artifact))
+
+                method = getattr(self.object.artifact, self.forward_method)
+                tmp = method(batch)
```

Comment:
a better name?

Reply:
This is already there, will address later.